src/RazorSlices/TextWriterHtmlExtensions.cs

```diff
@@ -89,4 +91,79 @@ private static bool TryHtmlEncodeAndWriteSpanFormattableSmall<T>(TextWriter text
         }
         return false;
     }
+
+#if NET8_0_OR_GREATER
+    public static void HtmlEncodeAndWriteUtf8SpanFormattable<T>(this TextWriter textWriter, T? formattable, HtmlEncoder htmlEncoder, ReadOnlySpan<char> format = default, IFormatProvider? formatProvider = null)
+        where T : IUtf8SpanFormattable
+    {
+        if (formattable is null)
+        {
+            return;
+        }
+
+        if (TryHtmlEncodeAndWriteUtf8SpanFormattableSmall(textWriter, formattable, htmlEncoder, format, formatProvider))
+        {
+            return;
+        }
+
+        var bufferSize = BufferSizes.SmallFormattableWriteByteSize * 2;
+        var rentedBuffer = ArrayPool<byte>.Shared.Rent(bufferSize);
+        int bytesWritten;
+
+        while (!formattable.TryFormat(rentedBuffer, out bytesWritten, format, formatProvider))
+        {
+            // Buffer was too small, return the current buffer and rent a new buffer twice the size
+            bufferSize = rentedBuffer.Length * 2;
+            ArrayPool<byte>.Shared.Return(rentedBuffer);
+            rentedBuffer = ArrayPool<byte>.Shared.Rent(bufferSize);
+        }
+
+        HtmlEncodeAndWriteUtf8(textWriter, rentedBuffer.AsSpan()[..bytesWritten], htmlEncoder);
+        ArrayPool<byte>.Shared.Return(rentedBuffer);
+    }
+
+    private static bool TryHtmlEncodeAndWriteUtf8SpanFormattableSmall<T>(TextWriter textWriter, T formattable, HtmlEncoder htmlEncoder, ReadOnlySpan<char> format = default, IFormatProvider? formatProvider = null)
+        where T : IUtf8SpanFormattable
+    {
+        Span<byte> buffer = stackalloc byte[BufferSizes.SmallFormattableWriteByteSize];
+        if (formattable.TryFormat(buffer, out var bytesWritten, format, formatProvider))
+        {
+            HtmlEncodeAndWriteUtf8(textWriter, buffer[..bytesWritten], htmlEncoder);
+            return true;
+        }
+        return false;
+    }
+#endif
+
+    public static void HtmlEncodeAndWrite(this TextWriter textWriter, ReadOnlySpan<char> value, HtmlEncoder htmlEncoder)
+    {
+        var buffer = ArrayPool<char>.Shared.Rent(value.Length);
+        value.CopyTo(buffer);
+        htmlEncoder.Encode(textWriter, buffer, 0, value.Length);
```

Comment:
Looks like an opportunity for a new API? Renting and copying is only needed because there is an API lacking for `ROS<char>`

Reply:
API like `TextEncoder.Encode(TextWriter, ReadOnlySpan<char>)`?

Reply:
Yeah

Reply:
@stephentoub thoughts?

Reply:
We could possibly add a `TextWriter, ReadOnlySpan<char>` overload, but note that the string/char[] overloads themselves just turn around and rent an array from the pool (they also need temporary space into which to encode before then writing that encoded output to the writer).

HtmlEncoder does have an Encode that takes a `ReadOnlySpan<char>`, it just writes to a destination span rather than to a TextWriter.

@grabyourpitchforks, did we consider adding a convenience span-based Encode with TextWriter?  I see one mentioned in https://github.com/dotnet/runtime/issues/27919 but that was closed.